apollo-router/src/query_planner/bridge_query_planner_pool.rs

```diff
@@ -0,0 +1,200 @@
+        schema: String,
+        configuration: Arc<Configuration>,
+        size: NonZeroUsize,
+    ) -> Result<Self, ServiceBuildError> {
+        let mut join_set = JoinSet::new();
+
+        let (sender, receiver) = bounded::<(
+            QueryPlannerRequest,
+            oneshot::Sender<Result<QueryPlannerResponse, QueryPlannerError>>,
+        )>(CHANNEL_SIZE);
+
+        let mut old_planners_iterator = old_planners.into_iter();
+
+        (0..size.into()).for_each(|_| {
+            let sdl = schema.clone();
+            let configuration = configuration.clone();
+
+            if let Some(old_planner) = old_planners_iterator.next() {
+                join_set.spawn(async move {
+                    BridgeQueryPlanner::new_from_planner(old_planner, sdl, configuration).await
+                });
+            } else {
+                join_set.spawn(async move { BridgeQueryPlanner::new(sdl, configuration).await });
+            }
+        });
+
+        let mut bridge_query_planners = Vec::new();
+
+        while let Some(task_result) = join_set.join_next().await {
+            let bridge_query_planner =
+                task_result.map_err(|e| ServiceBuildError::ServiceError(Box::new(e)))??;
+            bridge_query_planners.push(bridge_query_planner);
+        }
+
+        let schema = bridge_query_planners
+            .first()
+            .expect("There should be at least 1 service in pool")
+            .schema();
+
+        let subgraph_schemas = bridge_query_planners
+            .first()
+            .expect("There should be at least 1 service in pool")
+            .subgraph_schemas();
+
+        let planners = bridge_query_planners
+            .iter()
+            .map(|p| p.planner().clone())
+            .collect();
+
+        for (worker_id, mut planner) in bridge_query_planners.into_iter().enumerate() {
+            let receiver = receiver.clone();
+
+            tokio::spawn(async move {
+                while let Ok((request, res_sender)) = receiver.recv().await {
+                    let start = Instant::now();
+
+                    let svc = match planner.ready().await {
+                        Ok(svc) => svc,
+                        Err(e) => {
+                            if res_sender.send(Err(e)).is_err() {
+                                failfast_error!("receiver channel for query plan response was closed, this should never happen");
+                            }
+
+                            continue;
+                        }
+                    };
+
+                    let res = svc.call(request).await;
+
+                    f64_histogram!(
+                        "apollo.router.query_planner.duration",
+                        "Duration of the query planning.",
+                        start.elapsed().as_secs_f64(),
+                        [KeyValue::new("workerId", worker_id.to_string())]
+                    );
+
+                    if res_sender.send(res).is_err() {
```

Comment:
What happens here if the client has cancelled/gone away before the result is available?

Reply:
Right, we probably should. otoh as a followup i think we want to make sure a planned query always makes it back at least to the Caching service sitting above.

Reply:
Fixed in [1f7e478](https://github.com/apollographql/router/pull/4897/commits/1f7e478781f24f12f064ca669cddd8688b82db6d)